### ISSUE TITLE
insert_final_newline not respected

### ISSUE THREAD

**Issue by thet** [editorconfig/editorconfig-vim#136]
When editing a file with the following .editorconfig loaded:

[*]
indent_style = space
end_of_line = lf
insert_final_newline = true
trim_trailing_whitespace = true
charset = utf-8

no final newline is inserted.

**Comment by cxw42** [comment]
Thanks for reporting!  Would you please tell us the steps you took and how you discovered this?  For example, empty files should stay empty.  Could you share the file you were editing, as an attachment?  Much appreciated!

**Comment by keithbowes** [comment]
`:echo &eol` should give you 1.  You won't actually see a new line.

**Comment by istinspring** [comment]
hm, just trying to figure out why no newline inserted. 

similar config, neovim with editorconfig-vim.

**Comment by cxw42** [comment]
@istinspring Would you please tell us the steps you took that led you to discover the issue?  Please attach any files you used (but no private information, please).  Thanks!

**Comment by istinspring** [comment]
`
# EditorConfig helps developers define and maintain consistent
# coding styles between different editors and IDEs
# http://editorconfig.org

root = true

[*]

# Change these settings to your own preference
indent_style = space
indent_size = 2

# We recommend you to keep these unchanged
end_of_line = lf
charset = utf-8
trim_trailing_whitespace = true
insert_final_newline = true

[*.py]
indent_size = 4

[*.md]
trim_trailing_whitespace = false

[*.html]
indent_size = 2

[*.json]
indent_size = 2

[Makefile]
indent_style = tab
indent_size = 2
`
config. when you want to copy all lines in file you usually go `ggVG` in vim, but this last G do not cover last line.

**Comment by istinspring** [comment]
hm actually there is issue already https://github.com/editorconfig/editorconfig-vim/issues/74

so i used hexdump to look what's actually in file

```00001d0   e   t   i   m   e   ]       =       N   o   n   e  \n
00001de```

and looks like there is new line already? am i right?

**Comment by k-takata** [comment]
Vim assumes that every text line ends with a newline.
So, even if you don't see any empty line at the end of the file on Vim, the final newline will be added by default.
`insert_final_newline = true` is the default behavior of Vim.

If you set `insert_final_newline = false`, this plugin may remove the final newline by doing `setl nofixendofline`. (Actually, the final newline will be removed if the `eol` option is unset (`setl noeol`).)

**Comment by alexdilley** [comment]
> If you set `insert_final_newline = false`, this plugin may remove the final newline by doing `setl nofixendofline`. (Actually, the final newline will be removed if the `eol` option is unset (`setl noeol`).)

Yes, `setl nofixendofline` only honours the EOL of an existing file (but will add an `<EOL>` for a new file). `setl noeol` therefore feels more appropriate for a `insert_final_newline = false` config as it should be honoured in all cases, suggesting [this](https://github.com/editorconfig/editorconfig-vim/blob/047c4b455b9d56adee63a52d6d7cdf5bb6521d53/plugin/editorconfig.vim#L450) line is wrong?

**Comment by cxw42** [comment]
The specification is clear that `false` means "ensure [the file being saved] doesn’t" end with a newline.

From [the Vim docs](https://vimhelp.org/options.txt.html#%27endofline%27), it looks like we need to set both `noeol` and `nofixeol` to actually remove the newline.  PR welcome!

**Comment by Marlliton** [comment]
in my case I am trying not to insert a newline but the newline is always inserted at the end of the file.
root = true

```
root = true
[*]
indent_style = space
indent_size = 4
end_of_line = lf
charset = utf-8
trim_trailing_whitespace = true
[*.{ts,js}]
insert_final_newline = false
```

the code below is the same, the line is always inserted:

```
root = true
[*]
indent_style = space
indent_size = 4
end_of_line = lf
charset = utf-8
trim_trailing_whitespace = true

insert_final_newline = false
```

**Comment by TysonAndre** [comment]
> When insert_final_newline = true, setl fixendofline should be run. setl endofline could also be run, but this would have no effect unless the user manually changes fixendofline.

When I use `:r testname.out` to read the lines of a file (where testname.out has no trailing newline) into the end of a file or replace the end of a file (e.g. testname.phpt), the resulting file has no newline when saving without `setl endofline`.

So there are use cases for adding `endofline`. What I have this overridden to locally is:


```patch
--- a/plugin/editorconfig.vim
+++ b/plugin/editorconfig.vim
@@ -452,8 +452,10 @@ function! s:ApplyConfig(config) abort " Set the buffer options {{{1
         if exists('+fixendofline')
             if a:config["insert_final_newline"] == "false"
                 setl nofixendofline
+                setl noendofline
             else
                 setl fixendofline
+                setl endofline
             endif
```

**Comment by cxw42** [comment]
While reviewing this, I realized the specification is at odds with the behaviour being implemented in this PR.  I have opened https://github.com/editorconfig/editorconfig/issues/475 to clarify the specification.
